### ISSUE TITLE
单任务回归模型收敛困难，无法复现swift框架的效果

### ISSUE THREAD

**Issue by liuhongye216** [modelscope/ms-swift#7276]
我使用Swift框架训练单任务回归模型，取得了非常好的效果。但是当我自己复现swift框架中的回归代码时，训练过程中损失下降非常缓慢，最终验证集性能远低于预期，始终无法复现Swift现有框架在我的任务上展示的收敛效果。

输入: 化学分子字符串
输出: 连续数值（吸附量预测）
数据集规模:
训练集: ~80,000样本
验证集: ~10,000样本
测试集: ~10,000样本

- 基座模型: Llama3.1-8B-Instruct (本地模型)
- 模型路径: .../Model/3.1
- 模型类型: AutoModel + LayerNorm + Linear回归头
- 隐藏层维度: 4096
- 任务头结构: 
  - LayerNorm(4096)
  - Linear(4096 -> 1, bias=True)
- 池化策略: 取最后有效token的hidden_states

LoRA配置
task_type: SEQ_CLS
lora_rank: 8
lora_alpha: 32
lora_dropout: 0.05
target_modules: 
  - q_proj
  - v_proj
  - k_proj
  - o_proj
  - gate_proj
  - up_proj
  - down_proj

训练超参数
# 优化器配置
optimizer: AdamW
learning_rate: 1e-4
min_learning_rate: 1e-6
weight_decay: 0.1
adam_beta1: 0.9
adam_beta2: 0.95
adam_epsilon: 1e-8

# 学习率调度
lr_scheduler: cosine_with_warmup
warmup_ratio: 0.05  # 5% warmup
num_epochs: 10

# 批次配置
per_device_batch_size: 4
gradient_accumulation_steps: 4
effective_batch_size: 16  # 单GPU

# 梯度裁剪
max_grad_norm: 1.0

# 评估与保存
eval_steps: 500
save_steps: 500
logging_steps: 100

Tokenizer扩展
# 添加的特殊token
special_tokens = [
    '[PAD]',      # 专用padding token
    '[MASK]',     # 用于masking增强
    '[UNK_SMILES]'  # 无效SMILES的fallback
]

# SMILES化学token（完整元素周期表118个元素）
# 包括：
# - 裸露形式: H, C, N, O, Fe, Cu, Zn, ...
# - 方括号形式: [H], [C], [N], [Fe+2], [OH-], ...
# - 芳香形式: [c], [n], [o], ...
# - 立体化学: [@], [@@], [@H], [@@H]
# - 常见基团: C(=O)[O-], C#N, [N+](=O)[O-], ...

# 扩展后vocab大小: 128,256 → 约130,000+

用已有的框架训练的成功案例：

<img width="991" height="646" alt="Image" src="https://github.com/user-attachments/assets/be2314d2-db50-4106-a6a9-e9755673ee35" />

我的复现案例：
step,epoch,train_loss,grad_norm,learning_rate,timestamp
200,1,0.514309,52.497444,1.1737e-05,2026-01-04 02:05:47
400,1,3.412492,59.084221,2.3474e-05,2026-01-04 02:26:27
600,1,2.226220,37.895615,3.5211e-05,2026-01-04 02:46:59
800,1,0.839512,37.479275,4.6948e-05,2026-01-04 03:07:32
1000,1,0.337387,25.327797,5.8685e-05,2026-01-04 03:28:03
1200,1,3.181458,15.171706,7.0423e-05,2026-01-04 03:48:40
1400,1,0.647008,34.390358,8.2160e-05,2026-01-04 04:09:18
1600,1,0.079445,69.701332,9.3897e-05,2026-01-04 04:29:57
1800,1,2.003485,68.000694,9.9998e-05,2026-01-04 04:50:35
2000,1,0.712627,90.701347,9.9979e-05,2026-01-04 05:11:13
2200,1,1.598680,20.852528,9.9942e-05,2026-01-04 05:31:49
2400,1,0.783373,119.978554,9.9886e-05,2026-01-04 05:52:22
2600,1,1.000441,70.766464,9.9811e-05,2026-01-04 06:13:00
2800,1,0.659824,17.926046,9.9718e-05,2026-01-04 06:33:37
3000,1,2.196052,10.326244,9.9605e-05,2026-01-04 06:54:14
3200,1,2.032862,9.969392,9.9474e-05,2026-01-04 07:14:49
3400,1,4.752910,98.225143,9.9325e-05,2026-01-04 07:35:23
3600,2,2.546253,60.687935,9.9157e-05,2026-01-04 08:11:25
3800,2,1.215118,39.149750,9.8970e-05,2026-01-04 08:32:02
4000,2,0.685332,39.337929,9.8765e-05,2026-01-04 08:52:35
4200,2,1.262578,57.110092,9.8542e-05,2026-01-04 09:13:10
4400,2,0.996387,15.615241,9.8300e-05,2026-01-04 09:33:45
4600,2,1.505335,40.971756,9.8040e-05,2026-01-04 09:54:18
4800,2,0.890410,72.891907,9.7762e-05,2026-01-04 10:14:57
5000,2,1.186260,39.284813,9.7466e-05,2026-01-04 10:35:32
5200,2,2.936503,10.621700,9.7152e-05,2026-01-04 10:56:08
5400,2,2.246741,33.798607,9.6821e-05,2026-01-04 11:16:41
5600,2,0.856907,29.277584,9.6472e-05,2026-01-04 11:37:18
5800,2,2.394669,47.607555,9.6105e-05,2026-01-04 11:57:53
6000,2,2.335054,22.425270,9.5721e-05,2026-01-04 12:18:28
6200,2,1.034178,65.262581,9.5320e-05,2026-01-04 12:39:05
6400,2,1.356873,39.963478,9.4901e-05,2026-01-04 12:59:41
6600,2,0.141791,9.720350,9.4466e-05,2026-01-04 13:20:16
6800,2,2.261149,39.097504,9.4014e-05,2026-01-04 13:40:50
7000,3,0.990427,74.255180,9.3546e-05,2026-01-04 14:15:47
7200,3,1.372904,29.433332,9.3061e-05,2026-01-04 14:36:22
7400,3,0.665854,29.197678,9.2560e-05,2026-01-04 14:57:00
7600,3,0.671431,47.760815,9.2043e-05,2026-01-04 15:17:37
7800,3,0.824243,110.590500,9.1510e-05,2026-01-04 15:38:15
8000,3,0.277432,74.939644,9.0961e-05,2026-01-04 15:58:50
8200,3,1.013505,24.729574,9.0397e-05,2026-01-04 16:19:23
8400,3,2.465324,89.770332,8.9818e-05,2026-01-04 16:39:55
8600,3,4.244689,66.665306,8.9224e-05,2026-01-04 17:00:30
8800,3,1.623361,13.261815,8.8615e-05,2026-01-04 17:21:05
9000,3,1.066940,46.591156,8.7991e-05,2026-01-04 17:41:39
INFO:single_task_trainer:  Train Loss: 5.6785
INFO:single_task_trainer:  Val Loss: 1.3955
INFO:single_task_trainer:  Val MAE: 0.8830
INFO:single_task_trainer:  Val R²: 0.3530
我的训练损失下降缓慢，并且始终在波动，期望效果: R² > 0.8, RMSE < 0.6, MAE < 0.4

模型定义
class SwiftStyleSingleTaskModel(nn.Module):
    def __init__(self, base_model_path, num_targets=1, 
                 torch_dtype=torch.bfloat16, tokenizer=None):
        super().__init__()
        
        # 加载AutoModel
        self.model = AutoModel.from_pretrained(
            base_model_path,
            torch_dtype=torch_dtype,
            trust_remote_code=True,
            device_map='auto'
        )
        
        # 任务头
        hidden_size = self.model.config.hidden_size
        self.task_norm = nn.LayerNorm(hidden_size, dtype=torch_dtype)
        self.task_head = nn.Linear(hidden_size, num_targets, 
                                   bias=True, dtype=torch_dtype)
        
        # 初始化
        nn.init.normal_(self.task_head.weight, mean=0.0, std=0.02)
    
    def forward(self, input_ids, attention_mask=None, labels=None):
        # 获取hidden states
        outputs = self.model(input_ids=input_ids, 
                           attention_mask=attention_mask,
                           output_hidden_states=True)
        hidden_states = outputs.last_hidden_state
        
        # 池化到最后有效token
        sequence_lengths = self._get_sequence_lengths(input_ids, attention_mask)
        pooled = hidden_states[torch.arange(batch_size), sequence_lengths]
        
        # 预测
        features = self.task_norm(pooled)
        logits = self.task_head(features)
        
        # 计算损失
        loss = nn.functional.mse_loss(logits.float().squeeze(), 
                                     labels.float().squeeze())
        
        return {'loss': loss, 'logits': logits}

训练循环
# 优化器
optimizer = optim.AdamW(
    model.parameters(),
    lr=1e-4,
    betas=(0.9, 0.95),
    weight_decay=0.1,
    eps=1e-8
)

# 学习率调度
scheduler = get_cosine_schedule_with_warmup(
    optimizer,
    num_warmup_steps=int(total_steps * 0.05),
    num_training_steps=total_steps
)

# 训练循环
for epoch in range(10):
    for batch in train_loader:
        # 前向传播（BF16）
        with autocast(device_type='cuda', dtype=torch.bfloat16):
            outputs = model(**batch)
            loss = outputs['loss'] / gradient_accumulation_steps
        
        # 反向传播
        loss.backward()
        
        if (step + 1) % gradient_accumulation_steps == 0:
            # 梯度裁剪
            grad_norm = torch.nn.utils.clip_grad_norm_(
                model.parameters(), max_norm=1.0
            )
            
            # 更新参数
            optimizer.step()
            scheduler.step()
            optimizer.zero_grad()

我的训练问题是：模型架构不当；超参数配置问题；数据处理流程有误；还是训练策略需要调整？
Swift框架中针对类似任务的推荐配置是否有针对回归任务的特殊设置？

**Comment by Jintao-Huang** [comment]
没有特殊设置

self.task_norm = nn.LayerNorm(hidden_size, dtype=torch_dtype)

感觉是这个原因
